### ISSUE TITLE
Introduce a mechanism to assign and filter by test "tags"

### ISSUE THREAD

**Issue by famz** [avocado-framework/avocado#1368]
Tagging is a flexible way to manage a huge number of items. It is proposed to add a "tags" property to each `avocado.Test instance`, as a set of tag names, and allow querying and/or filtering during a test job.

For example I can have Test1 and Test2, where Test1 is tagged as "FeatureA, FeatureB", indicating that this test is focusing on execising features A and B; and Test2 is tagged as "FeatureA, FeatureC", which means it focuses on testing a different set of features.

Some time later I'm refactoring feature B, I'll want to know which tests are relevent to run (needless to say running all tests on all code changes is the less than most efficient way, with the test code growing). That's where tags come handy, and ideally I can throw all test cases into the job, and specify the tag I'm interested in:

```
avocado run Test1 Test2 --tags=FeatureB
```

then, Test2 is not run because it doesn't have the desired tag.

A tag can be applied either per source file or per instance.

**Comment by clebergnu** [comment]
Hi Fam,

The feature you suggested is great, and has indeed been in the back of our heads.  I decided to try it out, and have come up with a very rough draft of such a feature.  You can find it at:

https://github.com/clebergnu/avocado/tree/loader_tag_draft_1

Some information: because of the way Avocado inspects Python files to find tests (it doesn't load/run them until execution time), it's not easy to use Python decorators to individually set tags on test (methods).  The granularity we can achieve easily is a Python class, as you can see in the given example.

The usage is pretty simple, modeled after your suggestion.  To run all tests:

```
$ avocado run examples/tests/fast_and_slow.py
```

To run only fast tests:

```
$ avocado run examples/tests/fast_and_slow.py  --tags=fast
```

To run both slow and fast tests:

```
$ avocado run examples/tests/fast_and_slow.py  --tags=fast,slow
```

Please let us know if this matches your suggestion and use case.

Thanks!

**Comment by famz** [comment]
Looks pretty good, thanks! I have a question.

What is the syntax to assign multiple tags to a class? Like below?

    +class FastTest(Test):
    +
    +    """
    +    Fastest possible test
    +
    +    :avocado: tag=fast
    +    :avocado: tag=other
    +    """

**Comment by clebergnu** [comment]
@famz I guess I failed to support multiple tags.  I'll add proper support for that on the next iteration.

**Comment by clebergnu** [comment]
@famz I finally got back to work on this.  As mentioned earlier, two priorities here:

1) Make the test filter mechanism more generic
2) Support multiple test tags

My proposal is to add something like:

```
  --filter-by-tag TAGS  List of comma separated test tags to filter the tests
                        to be run. If multiple tags are given on a single
                        argument, test must have all tags. May be specified
                        multiple times to include more tests.
```

That is, one could run:

```
avocado run --filter-by-tags "fast,privileged" --filter-by-tags "slow,privileged" fast_and_slow.py
```

This would *not* run any of the tests on https://github.com/clebergnu/avocado/blob/loader_tag_draft_1/examples/tests/fast_and_slow.py since none of them have *both* fast and privileged tags.

To summarize, each `--filter-by-tags` would be an "or" operation, while the comma separated values in a given parameter would be an "and" operation.

How does it sound?

**Comment by famz** [comment]
Looks good to me. What is the syntax to specify multiple tags for each test class?

**Comment by clebergnu** [comment]
The idea is to follow you previous suggestion:

```
class FastTest(Test):

    """
    Fastest possible test

    :avocado: tag=fast
    :avocado: tag=privileged
    """
```

Unless we come up with a better idea.

**Comment by apahim** [comment]
@clebergnu, my first thought would be to run tests with *any* of the tags listed in the `--filter-by-tags`. So, `avocado run --filter-by-tags fast,privileged` would run the `FastTest()`, as it has one of the tags listed. I don't like this tag-grouping mechanism you described.

**Comment by clebergnu** [comment]
@apahim the idea is to allow for more flexible filtering.  If you want to run any, you'd just use: `--filter-by-tag fast --filter-by-tag privileged`.

I'm pretty sure requests for this kind of filter by exclusion will come up sooner or later, so I think we should be prepared.

**Comment by famz** [comment]
I was just speculating.. Can we save typing with something like this?

    :avocado: tags=fast, privileged

**Comment by apahim** [comment]
@clebergnu. I'm not convinced. For increased flexibility, I'd suggest to have a filter-out option (with higher priority), so one can run all tests, except the tests tagged with the tags listed in the filter-out.

**Comment by apahim** [comment]
@clebergnu,btw, I'm not saying the filter-out solves the same problem. From the user perspective, I just don't see the group-tagging as useful. But don't take this as granted... it might be use cases for it.

**Comment by clebergnu** [comment]
@apahim the "filter-out" you suggest is just the "and" operation, but applied separately.  It makes it impossible to specify some filtering scenarios though, since it will apply (per your suggestion) on *all* filters given.

**Comment by apahim** [comment]
@clebergnu sure, let's resume this subject in the pr so we can stress it out.

**Comment by adereis** [comment]
I find the mixing of OR and AND in the command line quite confusing... Maybe it's just me, but to be honest, I had to revisit your example more than once to understand what you wanted to accomplish.. :-(

I suggest a simpler mechanism, where tags are simply OR'ed (using one or multiple command line switches).  If you want to allow more complex filtering, then make it explicit, using extra command-line switches or some extra semantics.

For example:

  *  `--filter-by-tags=<list>`: only select tests that contain at least one of the tags from <list>.
  * `--filter-by-tags-required=<list>`: only select tests that contains all tags from <list>.
  * `--filter-by-tags-exclude=<list>`: exclude tests that contain any of the tags from <list>.

So your example would be created using `--filter-by-tags="fast,slow" --filter-by-tags-required="priviledged".

**Comment by clebergnu** [comment]
It's certainly not trivial to come up with a simple syntax that can express most filtering scenarios.  The proposed approach tried to be based on the simplest possible use case, that is, selecting tests that match *all* given tags.  I'd say that is the only rule you need to pay attention to start with.

To make it more visual, let's suppose the following tests have the following tags:

1) fast
2) slow
3) fast, privileged
4) slow, privileged
5) fast, filesystem
6) slow, filesystem

Now, if users want to select filesystem tests:

* `--filter-by-tags=filesystem` would return tests 5 and 6.

If users want fast **and** privileged tests, they would use:

* `--filter-by-tags=fast,privileged` would return only test 3.

IMHO, pretty simple.  If users want fast tests:

* `--filter-by-tags=fast` would return tests 1, 3 and 5.

This approach, without an explicit exclusion mechanism, wouldn't allow a user to select fast tests that are **not** privileged.  One possible solution to keep the single filtering rule unchaged, is to improve tagging on the tests, that is, the tagging could become:

1) fast, **unprivileged**
2) slow, **unprivileged**
3) fast, privileged
4) slow, privileged
5) fast, filesystem
6) slow, filesystem

Then we're back at the simple rule for choosing fast tests that are **not** privileged (that is, **unprivileged**):

* `--filter-by-tags=fast,unprivileged` would return test 1 only.

Now, that is still a single rule.  By allowing multiple occurrences of the same rule (the same command line option) complex selections can be made.  If users want to select, in addition to tests that are fast and unprivileged, tests that are filesystem related, they would use:

* `--filter-by-tags=fast,unprivileged --filter-by-tags=filesystem` would return tests 1, 5 and 6.

Maybe one usability improvement would be to use a `+` as a separator, to give the clear message that all given tags must exist on the test, that is:

* `--filter-by-tags=fast+unprivileged --filter-by-tags=filesystem` would return tests 1, 5 and 6.

Let me know if this sounds easier to understand.  Then, if that changes your perspective on the proposal.  Of course all approaches have ups and downs.  The downside of multiple command line options as you proposed, IMHO, is for instance, the ordering in which they get applied.  It may be just me, but I find it harder to draw a mental plan and to express that using those 3 options.

---

PS: This is somehow (reversly) inspired by the access rules mechanism of the Squid web proxy.  There, you'd have multiple rules ACLs statements that would be "ORed" together, say:

* acl internal_network src 192.168.1.0/24
* acl internal_network src 192.168.2.0/24

Then you could have one entry ACLs: 
* acl dmz_network dst 10.0.0.0/8
* acl webapp_ports tcp 443

To bind them (an "AND" operation) one would use:

* http_access internal_network dmz_network webapp_ports

Would allow source networks 192.168.1.0/24 and 192.168.2.0/24 to access the destination network 10.0.0.0/8 using the 443 tcp port only.

IMHO, it's simple enough, that is, you could start with a single entry ACLs and even single entry http_access statements.  Then you can build (almost) any scenario out of those 2 simple combinations.

**Comment by famz** [comment]
I think it makes sense to "OR" multiple `--filter-by-tags=` options, with "AND" tags inside each option. 

Additionally, we could have a "-" operator (meanwhile make it an illegal tag prefix) to negate individual tags, that way we can save the necessity of adding "unprivileged" tags, which IMO is very impractical.

To demostrate with the example by @clebergnu:

1) fast
2) slow
3) fast, privileged
4) slow, privileged
5) fast, filesystem
6) slow, filesystem

* `--filter-by-tags=filesystem` would return tests 5 and 6.
* `--filter-by-tags=filesystem --filter-by-tags=privileged`would return tests 3 to 6.
* `--filter-by-tags=filesystem,-slow --filter-by-tags=privileged`would return tests 3 to 5.
* `--filter-by-tags=filesystem,-slow --filter-by-tags=privileged,-fast`would return tests 4 and 5.

Does this make a flexible enough interface?

**Comment by clebergnu** [comment]
Yes, the idea of using a `-` to negate a given tag came during a discussion between myself and @apahim on IRC, and I think it would indeed increase the flexibility.  I was just afraid to introduce another option while the basic idea wasn't completely embraced though.

**Comment by ldoktor** [comment]
Hello guys, sorry for not commenting earlier. I recommended the use of `--filter-tags-only` and `--filter-tags-out`, which would extend the usability, but looking at the way it evolved I like the filter + `-` to restrict the `or`ed statements. Obviously combining all three would allow the greatest flexibility, but IMO it's not worth it.

Therefor from me I'd go with the `or`ed `and`s with the support to negate the term by `-` (basically what Fam described in the previous comment.

**Comment by adereis** [comment]
There are many ways to implement the tagging mechanism and sometimes it's a matter of taste. Since we're more or less converging on the proposal by Cleber and Fam, I suggest he turns his code into a proper PR to get it reviewed/merged.

**Comment by clebergnu** [comment]
@adereis yes, based on your previous verbal feedback, I'm already working on that.

**Comment by adereis** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [Experiment with filtering by test tags](https://trello.com/c/ulTkVqwv/804-experiment-with-filtering-by-test-tags)

**Comment by clebergnu** [comment]
@famz I have posted an update (PR #1621).  I'd love to hear your feedback on it.

**Comment by famz** [comment]
@clebergnu Thank you for implementing the tagging feature already! One remaining question I have is: should we document the best practices of tagging tests? So that it's more likely for independent test writers to tag their test cases, in a way that it's not only useful inside their own test repo, but also meaningful universely. I.e. Letting test authors know how to assign tags with a relatively common naming convention, as recommended by avocado.

Specifically, does it make sense to document on how (not) to use non-alphabetic characters and what is a good usage of upper/lower case, or suggest to use well known technical terms like "network", "cpu", "http", "backup" whenever possible, instead of trying to be distinctive like "my_network", "foo_cpu", "quick_http" or "fancy_backup"?

**Comment by clebergnu** [comment]
@famz that's a good point indeed.  I tried to come with some examples on the last commit of the series, but they're indeed far from a formal suggestion.

I've created a card so that we add this task to our planning:

https://trello.com/c/ZMmFqA5j/880-created-a-guideline-for-tagging-tests

If you don't mind, I'll close this issue, since the *main* task has been achieved and we now have a card for this new one.

Thanks!
